framework/src/org/checkerframework/framework/util/element/ElementAnnotationUtil.java

```diff
@@ -87,7 +87,17 @@ public static void applyAllElementAnnotations(
     static void addAnnotationsFromElement(
             final AnnotatedTypeMirror type, final List<? extends AnnotationMirror> annotations) {
         AnnotatedTypeMirror innerType = AnnotatedTypes.innerMostType(type);
-        innerType.addAnnotations(annotations);
+        if (innerType != type) {
+            for (AnnotationMirror annotation : annotations) {
+                if (innerType.addAnnotation(annotation)) {
+                    ErrorReporter.warnOnce(
+                            "Interpreting %s as a type annotation on an array component type.",
+                            annotation);
```

Comment:
You should be able to pass in the `Element` for the corresponding declaration to get a warning on that element, instead of a warning without line number.